demos/howtos/myula.py

```diff
@@ -0,0 +1,287 @@
+#
+# To illustrate MYULA, we will consider :math:`g(x) = \texttt{strength_reg} \  TV(x) = \texttt{strength_reg} \|\nabla x \|_{2, 1}`,
+# where :math:`\texttt{strength_reg}` is the regularization parameter which
+# controls the regularization strength induced by TV.
+
+# %%
+# Bayesian model definition
+# -------------------------
+# Then consider the following Bayesian Inverse Problem:
+#
+# .. math::
+#    \begin{align*}
+#    \mathbf{x} &\sim \exp (- \texttt{strength_reg} \|\nabla x \|_{2,1})\\
+#    \mathbf{y}_{obs} &\sim \mathcal{N}(\mathbf{A}\mathbf{x}, \texttt{sigma2}\,\mathbf{I}) \ ,
+#    \end{align*}
+#
+# with :math:`\texttt{sigma2}=0.05^2`.
+
+# %%
+# Likelihood definition
+# ---------------------
+# We first specify the data distribution as follows:
+sigma2=0.05**2
+y=cuqi.distribution.Gaussian(A, sigma2)
+# %%
+# Then we can define the likelihood with
+likelihood=y(y=y_obs)
+
+# %%
+# RestorationPrior and MoreauYoshidaPrior
+# ---------------------------------------
+# To apply MYULA, we need to define the Moreau-Yoshida prior 
+# :math:`\pi_{\texttt{smoothing_stength}}(x)`.
+# Evaluating this surrogate prior is doable but too intensive from
+# a computational point of view as it requires to solve an optimization problem.
+# However to apply MYULA, we only require access to
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{smoothing_strength}}`.
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{smoothing_strength}}`
+# is a denoising operator, which takes a signal as input and returns a less noisy
+# signal. In CUQIPy, we talk about restoration operators. It is an operator at the
+# core of a specific type of priors called RestorationPrior. We cannot sample from
+# these priors but they allow us to define other types of priors.
+
+#%%
+# RestorationPrior definition
+# ---------------------------
+# A restoration operator is associated with a parameter called
+# :math:`\texttt{restoration_strength}`. This parameter indicates how strong is
+# the restoration. For example, when this restoration is a denoiser, an operator
+# taking an signal as input and returning a less noisy signal, :math:`\texttt{restoration_strength}`
+# can correspond to the  denoising level.
+# In the following, we consider the denoising restoration operator
+# :math:`\operatorname{prox}_{\texttt{strength_reg}\ TV}^{\texttt{restoration_strength}}`.
+# We use the implementation provided by Scikit-Image. But we can use any solver
+# to compute this quantity.
+# We emphasize that we have for any :math:`g`
+#
+# .. math::
+#       \operatorname{prox}_{\texttt{strength_reg}\  g}^{\texttt{smoothing_strength}} = \operatorname{prox}_{g}^{\texttt{weight}} ,
+#
+# with :math:`\texttt{weight} = \texttt{strength_reg} \times  \texttt{smoothing_strength}`.
+strength_reg = 10
+restoration_strength = 0.5*sigma2
+from skimage.restoration import denoise_tv_chambolle
+def prox_g(x, strength_reg=None, restoration_strength=None):
```

Comment:
I don't understand the role of strength_reg here. Is it needed? Or could it be omitted such that prox_g only accepts x and restoration_strength as inputs? I think that would simplify the following code, since the the kwargs would not be needed?

Reply:
That is a good point.
First, we changed the name strength_reg to regularization_strength in order to be a bit more explicit.
regularization_strength corresponds to the regularization parameter of the total-variation. It is needed to control the impact of the prior (or regularization). Here we don't smooth the total variation TV only but the regularizer = the negative log-prior = regularization_strength x TV.